include/sampleflow/consumers/autocovariance_dim_1.h

```diff
@@ -0,0 +1,255 @@
+         * The current value of $\bar x_k$ as described in the introduction
+         * of this class.
+         */
+
+        InputType           current_mean;
+
+        /**
+         * Currently describes lag of autocovariance
+         */
+
+        InputType 			k;
+
+        /**
+         * The current value of $\bar x_k$ as described in the introduction
+         * of this class.
+         */
+        boost::numeric::ublas::matrix<scalar_type> current_autocovariationdim1_matrix;
+
+        /**
+        * Save previous sample value
+        */
+
+        boost::numeric::ublas::matrix<scalar_type> past_sample;
+        boost::numeric::ublas::matrix<scalar_type> past_sample_replace;
+
+        /**
+         * The number of samples processed so far.
+         */
+        types::sample_index n_samples;
+    };
+
+    template <typename InputType>
+    autocovariancedim1 <InputType>::
+	autocovariancedim1  ()
+      :
+      n_samples (0)
+    {}
+
+
+    template <typename InputType>
+    void
+	autocovariancedim1 <InputType>::
+    consume (InputType sample, AuxiliaryData /*aux_data*/)
+    {
+      std::lock_guard<std::mutex> lock(mutex);
+
+      // If this is the first sample we see, initialize the matrix with
+      // this sample. After the first sample, the autocovariance matrix
+      // is the zero matrix since a single sample does not have any friends yet.
+
+      //Set how long is our autocovariance tail
+      k=10;
+
+      if (n_samples == 0)
+        {
+          n_samples = 1;
+          current_autocovariationdim1_matrix.resize (k,3);
+          for (unsigned int i=0; i<k; ++i)
+        	for (unsigned int j=0; j<2; ++j){
+        		current_autocovariationdim1_matrix(i,j) = 0;
+        	}
+          current_mean = sample;
+          past_sample.resize (k,1);
+          past_sample_replace.resize (k,1);
+          past_sample(0,0) = sample;
+        }
+      else if(n_samples < k)
+		{
+     	 ++n_samples;
+    	for (unsigned int i=0; i<n_samples-1; ++i){
+    		//Update first value (alpha)
+    		    InputType alpha = sample;
+    		    alpha *= past_sample(i,0);
+    		    alpha -= current_autocovariationdim1_matrix(i,0);
+    		    alpha /= n_samples;
+    		    current_autocovariationdim1_matrix(i,0) += alpha;
+
+    		//Update second value (beta)
+    		    InputType beta = sample;
+    			beta += past_sample(i,0);
+    			beta -= current_autocovariationdim1_matrix(i,1);
+    			beta /= n_samples;
+    			current_autocovariationdim1_matrix(i,1) += beta;
+    		}
+
+    		//Save needed past values
+    	for (unsigned int i=0; i<n_samples-1; ++i){
+    		past_sample_replace(i+1,0)=past_sample(i,0);
+    		}
+    	past_sample_replace(0,0) = sample;
+    	past_sample = past_sample_replace;
+
+    	// Then also update the running mean:
+    	 InputType update = sample;
+    	          update -= current_mean;
+    	          update /= n_samples;
+    	          current_mean += update;
+		}
+     else
+     {
+        ++n_samples;
```

Comment:
You get into this `else` branch if `n_samples>=k`. In that case, you probably just want to throw away the first sample you stored. I think you do that below in line 220, but there are easier ways. For example, using the `std::list` or `std::dequeue` data structures, there are `push_back()` and `pop_front()` functions that allow you to do exactly what you want to do. We can talk about this on Monday.